src/interp/interp.cc

```diff
@@ -245,9 +245,17 @@ void Store::DeleteRoot(RootList::Index index) {
 }
 
 void Store::Collect() {
+  assert(gc_context_ == nullptr);
+
   size_t object_count = objects_.size();
-  marks_.resize(object_count);
-  std::fill(marks_.begin(), marks_.end(), false);
+
+  GCContext context;
+
+  context.call_depth = 0;
+  context.marks.resize(object_count);
+  std::fill(context.marks.begin(), context.marks.end(), false);
```

Comment:
Is this line needed?  Wont the resize above take care of initializing to false?  (https://stackoverflow.com/questions/22983707/stdvectorbool-guaranteed-to-default-all-entries-to-false/22984114)

Reply:
It was in the original code. I can remove it of course.

Reply:
Right, but in the original code the `marks_` object persisted between calls to Collect.   In this new code `context.marks` is new on each call. 

Speaking of which would it make sense to keep the `GCContext` in the Store rather than creating it anew each `Collect`?

Reply:
Could be, but that increases memory consumption.

I have another suggestion: what about making this class public as a member of `Store`. The name could be `MarkContext` or something else. Then move the `Mark(object)` function from `Store` to this new class, and change the `Mark()` virtual function of Object to get this class as an argument rather than the `Store`. The problem is, you can call the `Mark()` function anytime, which is incorrect if you are not in the `Collect()` function. This way the compiler protects you from misusing `Mark()`, because it simply does not exist outside of `Collect()` function.

Reply:
What do you mean "increases memory consumption." ?  With the old code the "marks_" object was always allocated as part of the store and always as large as the number of objects from the most recent collection.     Storing GCContext in directly in the store seems pretty much exactly the same size.

I also think its a reasonable to have the `GCContext` persist otherwise it will just need to get relocated on the heap constantly.

Reply:
I know the old code always used some extra memory, and I wanted to reduce it. But we can keep the current approach of course, and I think it is not worth to introduce the `GCContext` in this case. The advantage of the current approach is that if the object count does not change, the reallocation is basically free. Ok, I keep the current approach.